File: gl4java-impl/ExaminerViewer.java
/*
 * gleem -- OpenGL Extremely Easy-To-Use Manipulators.
 * Copyright (C) 1998, 1999, 2002 Kenneth B. Russell (kbrussel@alum.mit.edu)
 * See the file LICENSE.txt in the doc/ directory for licensing terms.
 */

package gnu.gleem;

import java.awt.Dimension;
import java.awt.event.*;
import java.util.*;

import gnu.gleem.linalg.*;
import gl4java.*;
import gl4java.drawable.*;

/** <P> This is an application-level class, not part of the
    manipulator hierarchy. It is an example of how you might integrate
    gleem with another application which uses the mouse. </P>

    <P> For the given GLDrawable, the ExaminerViewer takes over the
    setting of the view position. It passes along mouse events it is
    not interested in to the ManipManager's mouse routines. </P>

    <P> The ExaminerViewer's controls are similar to those of Open
    Inventor's Examiner Viewer. Alt + Left mouse button causes
    rotation about the focal point. Alt + Right mouse button causes
    translation parallel to the image plane. Alt + both mouse buttons,
    combined with up/down mouse motion, causes zooming out and in
    along the view vector. (On platforms with a "Meta" key, that key
    can be substituted in place of the Alt key.) </P>

    <P>NOTE: the current ExaminerViewer implementation assumes a
    minimum of two mouse buttons. For the Mac OS, the code needs to be
    adjusted to use e.g., the Control key as the "right" mouse
    button. </P> */

public class ExaminerViewer {
  private GLDrawable window;
  /** Simple state machine for figuring out whether we are grabbing
      events */
  private boolean interactionUnderway;
  private boolean iOwnInteraction;

  /** Simple state machine for computing distance dragged */
  private boolean button1Down;
  private boolean button2Down;
  private int numMouseButtons;
  private int lastX;
  private int lastY;

  /** Camera parameters */
  Vec3f position;
  Rotf orientation;
  float focalDist;
  float minFocalDist;
  float rotateSpeed;
  float minRotateSpeed;
  float dollySpeed;
  float minDollySpeed;
  CameraParameters params;

  /** Our bounding sphere provider (for viewAll()) */
  BSphereProvider provider;

  private MouseMotionAdapter mouseMotionListener = new MouseMotionAdapter() {
      public void mouseDragged(MouseEvent e) {
        motionMethod(e, e.getX(), e.getY());
      }

      public void mouseMoved(MouseEvent e) {
        passiveMotionMethod(e);
      }
    };

  private MouseAdapter mouseListener = new MouseAdapter() {
      public void mousePressed(MouseEvent e) {
        mouseMethod(e, e.getModifiers(), true, e.getX(), e.getY());
      }

      public void mouseReleased(MouseEvent e) {
        mouseMethod(e, e.getModifiers(), false, e.getX(), e.getY());
      }
    };

  private GLEventListener glListener = new GLEventListener() {
      public void init(GLDrawable drawable) {}
      public void preDisplay(GLDrawable drawable) {}
      public void display(GLDrawable drawable) {}
      public void postDisplay(GLDrawable drawable) {}
      public void cleanup(GLDrawable drawable) {
        detach();
      }
      public void reshape(GLDrawable drawable, int width, int height) {
        reshapeMethod(width, height);
      }
    };

  /** The constructor takes the number of mouse buttons on this system
      (couldn't figure out how to determine this internally) */
  public ExaminerViewer(int numMouseButtons) {
    this.numMouseButtons = numMouseButtons;

    focalDist = 10.0f;
    minFocalDist = 1.0f;
    rotateSpeed = 4.0f;
    minRotateSpeed = 0.0001f;
    dollySpeed = 2.0f;
    minDollySpeed = 0.0001f;
    position = new Vec3f();
    orientation = new Rotf();
    params = new CameraParameters();
  }

  /** <P> Attaches this ExaminerViewer to the given GLDrawable. This
      causes the ManipManager's mouse routines to be removed from the
      window (using ManipManager.removeMouseListeners) and the
      ExaminerViewer's to be installed. The GLDrawable should be
      registered with the ManipManager before the ExaminerViewer is
      attached to it. </P>

      <P> In order for the viewer to do anything useful, you need to
      provide a BSphereProvider to it to allow "view all"
      functionality. </P> */
  public void attach(GLDrawable window, BSphereProvider provider) {
    this.window = window;
    this.provider = provider;
    init();
    setupListeners();
  }

  /** Detaches from the given window. This causes the ManipManager's
      mouse listeners to be reinstalled on the GLDrawable and the
      ExaminerViewer's to be removed. */
  public void detach() {
    removeListeners();
    this.window = null;
    this.provider = null;
  }
  
  /** Call this at the end of your display() method to cause the
      Modelview matrix to be recomputed for the next frame. */
  public void update(GLFunc gl, GLUFunc glu) {
    recalc(gl, glu);
  }

  /** Call this from within your display() method to cause the
      ExaminerViewer to recompute its position based on the visible
      geometry. A BSphereProvider must have already been set or this
      method has no effect. */
  public void viewAll(GLFunc gl, GLUFunc glu) {
    if (provider == null) {
      return;
    }
    // Figure out how far to move
    float vertFOV, horizFOV, minFOV;
    vertFOV = 2.0f * params.getVertFOV();
    horizFOV = 2.0f * (float) Math.atan(params.getImagePlaneAspectRatio() *
                                        Math.tan(params.getVertFOV()));
    if (vertFOV < horizFOV)
      minFOV = vertFOV;
    else
      minFOV = horizFOV;
    if (minFOV == 0.0f) {
      throw new RuntimeException("Minimum field of view was zero");
    }
    BSphere bsph = provider.getBoundingSphere();
    float dist = bsph.getRadius() / (float) Math.sin(minFOV / 2.0f);
  
    // Now position the camera this far back from the scene's center
    // along the negative of the current forward direction
    Vec3f dir = orientation.rotateVector(Vec3f.Z_AXIS);
    dir.scale(dist / dir.length());
    position.add(dir, bsph.getCenter());
    focalDist = dist;
    recalc(gl, glu);
  }

  /** Get the camera parameters out of this Examiner Viewer (for
      example, to pass to ManipManager.updateCameraParameters()) */
  public CameraParameters getCameraParameters() {
    return params;
  }

  /** These routines can be hooked into a GUI by calling them from
      ActionEvent listeners for buttons elsewhere in the application. */
  public void rotateFaster() {
    rotateSpeed *= 2.0f;
  }

  public void rotateSlower() {
    if (rotateSpeed < minRotateSpeed)
      return;
    else
      rotateSpeed /= 2.0f;
  }

  public void dollyFaster() {
    dollySpeed *= 2.0f;
  }

  public void dollySlower() {
    if (dollySpeed < minDollySpeed)
      return;
    else
      dollySpeed /= 2.0f;
  }

  //----------------------------------------------------------------------
  // Internals only below this point
  //

  private static final float EPSILON = 0.0001f;

  private void setupListeners() {
    ManipManager.getManipManager().removeMouseListeners(window);
    window.addMouseMotionListener(mouseMotionListener);
    window.addMouseListener(mouseListener);
    window.addGLEventListener(glListener);
  }

  private void removeListeners() {
    window.removeMouseMotionListener(mouseMotionListener);
    window.removeMouseListener(mouseListener);
    window.removeGLEventListener(glListener);
    ManipManager.getManipManager().setupMouseListeners(window);
  }

  private void passiveMotionMethod(MouseEvent e) {
    ManipManager.getManipManager().mouseMoved(e);
  }

  private static boolean modifiersMatch(MouseEvent e) {
    return ((e.isAltDown() || e.isMetaDown()) &&
            (!e.isControlDown() && !e.isShiftDown()));
  }

  private void init() {
    focalDist = 10.0f;
    minFocalDist = 1.0f;
    rotateSpeed = 4.0f;
    minRotateSpeed = 0.0001f;
    dollySpeed = 2.0f;
    minDollySpeed = 0.0001f;
    interactionUnderway = false;
    iOwnInteraction = false;
    button1Down = false;
    button2Down = false;
    provider = null;

    position.set(0, 0, 0);

    Dimension size = window.getSize();
    int xSize = size.width;
    int ySize = size.height;
    params.setPosition(position);
    params.setForwardDirection(Vec3f.NEG_Z_AXIS);
    params.setUpDirection(Vec3f.Y_AXIS);
    params.setVertFOV((float) Math.PI / 8.0f);
    params.setImagePlaneAspectRatio((float) xSize / (float) ySize);
    params.setXSize(xSize);
    params.setYSize(ySize);
  }

  private void motionMethod(MouseEvent e, int x, int y) {
    if (interactionUnderway && !iOwnInteraction) {
      ManipManager.getManipManager().mouseDragged(e);
    } else {
      int dx = x - lastX;
      int dy = y - lastY;
  
      lastX = x;
      lastY = y;

      if ((button1Down && (!button2Down))) {

        // Rotation functionality
        float xRads = (float) Math.PI * -1.0f * dy * rotateSpeed / 1000.0f;
        float yRads = (float) Math.PI * -1.0f * dx * rotateSpeed / 1000.0f;
        Vec3f fw = new Vec3f(0, 0, -1);
        Vec3f diff = new Vec3f(orientation.rotateVector(fw));
        diff.normalize();
        diff.scale(focalDist);
        // Update by this rotation
        Rotf xRot = new Rotf(Vec3f.X_AXIS, xRads);
        Rotf yRot = new Rotf(Vec3f.Y_AXIS, yRads);
        Rotf newRot = yRot.times(xRot);
        orientation = orientation.times(newRot);
        // Now update the camera's position. Rotate about the focal point.
        Vec3f backOut = orientation.rotateVector(fw);
        backOut.normalize();
        backOut.scale(-1.0f * focalDist);
        position = position.plus(backOut).plus(diff);

      } else if (button2Down && (!button1Down)) {

        // Translate functionality
        // Compute the local coordinate system's difference vector
        Vec3f localDiff = new Vec3f(dollySpeed * -1.0f * dx / 100.0f,
                                    dollySpeed * dy / 100.0f,
                                    0.0f);
        // Rotate this by camera's orientation
        Vec3f worldDiff = orientation.rotateVector(localDiff);
        // Add on to position
        position.add(worldDiff);

      } else if (button1Down && button2Down) {

        // Dolly functionality
        // Compute the local coordinate system's difference vector
        Vec3f localDiff = new Vec3f(0, 0, dollySpeed * -1.0f * dy / 100.0f);
        // Rotate this by camera's orientation
        Vec3f worldDiff = orientation.rotateVector(localDiff);
        position.add(worldDiff);
        // Subtract off the dolly amount from the focal distance
        float focalDiff = (float) worldDiff.length();
        if (dy > 0)
          focalDiff *= -1.0f;
        focalDist += focalDiff;
        if (focalDist < minFocalDist)
          focalDist = minFocalDist;
      }

      // Force redraw if window will not do it automatically
      if (!(window instanceof GLRunnable)) {
        window.repaint();
      }
    }

    // FIXME: can not call this from here
    //    recalc();
  }

  private void mouseMethod(MouseEvent e, int mods, boolean press,
                           int x, int y) {
    if ((interactionUnderway && !iOwnInteraction) ||
        (!modifiersMatch(e))) {
      // Update state and pass this event along to the ManipManager
      if (press) {
        interactionUnderway = true;
        iOwnInteraction = false;
        ManipManager.getManipManager().mousePressed(e);
      } else {
        interactionUnderway = false;
        iOwnInteraction = false;
        ManipManager.getManipManager().mouseReleased(e);
      }
    } else {
      if ((mods & MouseEvent.BUTTON1_MASK) != 0) {
        if (press) {
          button1Down = true;
        } else {
          button1Down = false;
        }
      } else {
        if (numMouseButtons == 2) {
          if ((mods & MouseEvent.BUTTON2_MASK) != 0) {
            if (press) {
              button2Down = true;
            } else {
              button2Down = false;
            }
          }
        } else if (numMouseButtons == 3) {
          // FIXME: must test this on 3-button system
          if ((mods & MouseEvent.BUTTON3_MASK) != 0) {
            if (press) {
              button2Down = true;
            } else {
              button2Down = false;
            }
          }
        }
      }

      lastX = x;
      lastY = y;

      if (button1Down || button2Down) {
        interactionUnderway = true;
        iOwnInteraction = true;
      } else {
        interactionUnderway = false;
        iOwnInteraction = false;
      }

      // Force redraw if window will not do it automatically
      if (!(window instanceof GLRunnable)) {
        window.repaint();
      }
    }
  }

  private void reshapeMethod(int w, int h) {
    float aspect, theta;
    aspect = (float) w / (float) h;
    if (w >= h)
      theta = 45;
    else
      theta = (float) Math.toDegrees(Math.atan(1 / aspect));
    params.setVertFOV((float) (Math.toRadians(theta) / 2.0));
    params.setImagePlaneAspectRatio(aspect);
    params.setXSize(w);
    params.setYSize(h);
  }

  private void recalc(GLFunc gl, GLUFunc glu) {
    // Recompute position, forward and up vectors
    params.setPosition(position);
    Vec3f tmp = new Vec3f();
    orientation.rotateVector(Vec3f.NEG_Z_AXIS, tmp);
    params.setForwardDirection(tmp);
    orientation.rotateVector(Vec3f.Y_AXIS, tmp);
    params.setUpDirection(tmp);

    // Compute modelview matrix based on camera parameters, position and
    // orientation
    gl.glMatrixMode(GLEnum.GL_MODELVIEW);
    gl.glLoadIdentity();
    float ang = orientation.get(tmp);
    if (tmp.lengthSquared() > EPSILON)
      gl.glRotatef((float) Math.toDegrees(ang), -tmp.x(), -tmp.y(), -tmp.z());
    gl.glTranslatef(-position.x(), -position.y(), -position.z());
    gl.glMatrixMode(GLEnum.GL_PROJECTION);
    gl.glLoadIdentity();
    glu.gluPerspective(Math.toDegrees(params.getVertFOV() * 2.0),
                       params.getImagePlaneAspectRatio(),
                       1, 100);
  }
}

File: src/ExaminerViewer.cpp
/*
 * gleem -- OpenGL Extremely Easy-To-Use Manipulators.
 * Copyright (C) 1998 Kenneth B. Russell (kbrussel@media.mit.edu)
 * See the file LICENSE.txt in the doc/ directory for licensing terms.
 */

#include <math.h>
#include <stdlib.h>
#include <iostream>
#include <assert.h>
#ifdef WIN32
# include <windows.h>
#endif
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>
#include <ExaminerViewer.h>
#include <MathUtil.h>

GLEEM_USE_NAMESPACE

#define EPSILON 0.0001f

// How many examiner viewers do we expect to instantiate?
static const int NUM_EXAMINER_VIEWERS = 16;
ExaminerViewer::WindowToExaminerViewerTable
ExaminerViewer::windowViewerTable(NUM_EXAMINER_VIEWERS, hash<int>());

ExaminerViewer::ExaminerViewer()
{
  glutInitWindowSize(256, 256);
  init(glutCreateWindow("Examiner Viewer"));
  pair<WindowToExaminerViewerTable::iterator, bool> result =
    windowViewerTable.insert_unique(windowID, this);
  assert(result.second == true);
}

ExaminerViewer::~ExaminerViewer()
{
  windowViewerTable.erase(windowViewerTable.find(windowID));
  glutDestroyWindow(windowID);
}

ExaminerViewer::ExaminerViewer(const ExaminerViewer &viewer)
{
  glutInitWindowSize(256, 256);
  init(glutCreateWindow("Examiner Viewer"));
  *this = viewer;
}

ExaminerViewer &
ExaminerViewer::operator=(const ExaminerViewer &viewer)
{
  mouseDelegate = viewer.mouseDelegate;
  motionDelegate = viewer.motionDelegate;
  position = viewer.position;
  orientation = viewer.orientation;
  focalDist = viewer.focalDist;
  minFocalDist = viewer.minFocalDist;
  rotateSpeed = viewer.rotateSpeed;
  minRotateSpeed = viewer.minRotateSpeed;
  dollySpeed = viewer.dollySpeed;
  minDollySpeed = viewer.minDollySpeed;
  params = viewer.params;
  interactionUnderway = false;
  iOwnInteraction = false;
  provider = viewer.provider;
  return *this;
}

void
ExaminerViewer::update()
{
  recalc();
}

const CameraParameters &
ExaminerViewer::getCameraParameters() const
{
  return params;
}

const GleemRot &
ExaminerViewer::getOrientation() const
{
  return orientation;
}

void
ExaminerViewer::setMouseDelegate(void (*mouseDelegate)(int button, int state,
						       int x, int y))
{
  this->mouseDelegate = mouseDelegate;
}

void
ExaminerViewer::setMotionDelegate(void (*motionDelegate)(int x, int y))
{
  this->motionDelegate = motionDelegate;
}

void
ExaminerViewer::setBSphereProvider(BSphereProvider *provider)
{
  this->provider = provider;
}

void
ExaminerViewer::viewAll()
{
  if (provider == NULL)
    {
      cerr << "gleem::ExaminerViewer::viewAll: ERROR: "
	   << "bounding sphere provider was NULL, so I don't know how "
	   << "big the scene is" << endl;
      return;
    }
  // Figure out how far to move
  float vertFOV, horizFOV, minFOV;
  vertFOV = 2.0f * params.vertFOV;
  horizFOV = 2.0f * atan(params.imagePlaneAspectRatio * tan(params.vertFOV));
  if (vertFOV < horizFOV)
    minFOV = vertFOV;
  else
    minFOV = horizFOV;
  if (minFOV == 0.0f)
    {
      cerr << "gleem::ExaminerViewer::viewAll: ERROR: "
	   << "minimum field of view was zero!" << endl;
      return;
    }
  const BSphere &bsph = provider->getBoundingSphere();
  float dist = bsph.getRadius() / (float) sin(minFOV / 2.0f);
  
  // Now position the camera this far back from the scene's center
  // along the negative of the current forward direction
  GleemV3f bw(0.0f, 0.0f, 1.0f);
  GleemV3f dir = orientation.rotateVector(bw);
  GleemV3f::scale(dir, dist / dir.length(), dir);
  GleemV3f::add(dir, bsph.getCenter(), position);
  focalDist = dist;
  recalc();
}

void
ExaminerViewer::rotateFaster()
{
  rotateSpeed *= 2.0f;
}

void
ExaminerViewer::rotateSlower()
{
  if (rotateSpeed < minRotateSpeed)
    return;
  else
    rotateSpeed /= 2.0f;
}

void
ExaminerViewer::dollyFaster()
{
  dollySpeed *= 2.0f;
}

void
ExaminerViewer::dollySlower()
{
  if (dollySpeed < minDollySpeed)
    return;
  else
    dollySpeed /= 2.0f;
}

void
ExaminerViewer::setTitle(const char *title)
{
  int curWindow = glutGetWindow();
  if (curWindow != windowID)
    {
      glutSetWindow(windowID);
      glutSetWindowTitle((char *) title);
      glutSetWindow(curWindow);
    }
  else
    glutSetWindowTitle((char *) title);
}

void
ExaminerViewer::makeCurrent()
{
  glutSetWindow(windowID);
}

int
ExaminerViewer::getWindowID() const
{
  return windowID;
}

void
ExaminerViewer::init(int windowID)
{
  this->windowID = windowID;
  glutSetWindow(windowID);

  mouseDelegate = NULL;
  motionDelegate = NULL;

  focalDist = 10.0f;
  minFocalDist = 1.0f;
  rotateSpeed = 4.0f;
  minRotateSpeed = 0.0001f;
  dollySpeed = 2.0f;
  minDollySpeed = 0.0001f;
  interactionUnderway = false;
  iOwnInteraction = false;
  button1Down = false;
  button2Down = false;
  provider = NULL;

  position.setValue(0, 0, 0);

  params.position.setValue(0, 0, 0);
  params.forwardDirection.setValue(0, 0, -1);
  params.upDirection.setValue(0, 1, 0);
  params.vertFOV = M_PI / 8.0;
  params.imagePlaneAspectRatio = 1;
  params.xSize = 256;
  params.ySize = 256;

  glutMouseFunc(mouseFunc);
  glutMotionFunc(motionFunc);
  glutReshapeFunc(reshapeFunc);
  numMouseButtons = glutDeviceGet(GLUT_NUM_MOUSE_BUTTONS);
}

void
ExaminerViewer::mouseFunc(int button, int state, int x, int y)
{
  getExaminerViewer(glutGetWindow())->mouseMethod(button, state, x, y);
}

void
ExaminerViewer::motionFunc(int x, int y)
{
  getExaminerViewer(glutGetWindow())->motionMethod(x, y);
}

void
ExaminerViewer::reshapeFunc(int w, int h)
{
  getExaminerViewer(glutGetWindow())->reshapeMethod(w, h);
}

void
ExaminerViewer::mouseMethod(int button, int state, int x, int y)
{
  int mods = glutGetModifiers();
  if ((interactionUnderway && !iOwnInteraction) ||
      (!(mods & GLUT_ACTIVE_ALT)))
    {
      if (state == GLUT_DOWN)
	{
	  interactionUnderway = true;
	  iOwnInteraction = false;
	}
      else
	{
	  interactionUnderway = false;
	  iOwnInteraction = false;
	}
      // Pass this event on to the mouse delegate
      if (mouseDelegate != NULL)
	(*mouseDelegate)(button, state, x, y);
    }
  else
    {
      if (button == GLUT_LEFT_BUTTON)
	if (state == GLUT_DOWN)
	  button1Down = true;
	else
	  button1Down = false;

      if (numMouseButtons == 2)
	{
	  if (button == GLUT_RIGHT_BUTTON)
	    if (state == GLUT_DOWN)
	      button2Down = true;
	    else
	      button2Down = false;
	}
      else if (numMouseButtons == 3)
	{
	  if (button == GLUT_MIDDLE_BUTTON)
	    if (state == GLUT_DOWN)
	      button2Down = true;
	    else
	      button2Down = false;
	}
      lastX = x;
      lastY = y;
      
      if (button1Down || button2Down)
	{
	  interactionUnderway = true;
	  iOwnInteraction = true;
	}
      else
	{
	  interactionUnderway = false;
	  iOwnInteraction = false;
	}
    }
}

void
ExaminerViewer::motionMethod(int x, int y)
{
  if (interactionUnderway && !iOwnInteraction)
    {
      // Pass this event on to the motion delegate
      if (motionDelegate != NULL)
	(*motionDelegate)(x, y);
    }
  else 
    {
      int dx = x - lastX;
      int dy = y - lastY;
  
      lastX = x;
      lastY = y;

      if ((button1Down && (!button2Down)))
	{
	  // Rotation functionality
	  float xRads = M_PI * -1.0f * dy * rotateSpeed / 1000.0f;
	  float yRads = M_PI * -1.0f * dx * rotateSpeed / 1000.0f;
	  GleemV3f fw(0, 0, -1);
	  GleemV3f diff = orientation.rotateVector(fw);
	  diff.normalize();
	  GleemV3f::scale(diff, focalDist, diff);
	  // Update by this rotation
	  GleemRot xRot(GleemV3f(1, 0, 0), xRads);
	  GleemRot yRot(GleemV3f(0, 1, 0), yRads);
	  GleemRot newRot = yRot * xRot;
	  orientation = orientation * newRot;
	  // Now update the camera's position. Rotate about the focal point.
	  GleemV3f backOut = orientation.rotateVector(fw);
	  backOut.normalize();
	  GleemV3f::scale(backOut, -1.0f * focalDist, backOut);
	  position = position + backOut + diff;
	}
      else if (button2Down && (!button1Down))
	{
	  // Translate functionality
	  // Compute the local coordinate system's difference vector
	  GleemV3f localDiff(dollySpeed * -1.0f * dx / 100.0f,
			    dollySpeed * dy / 100.0f,
			    0.0f);
	  // Rotate this by camera's orientation
	  GleemV3f worldDiff = orientation.rotateVector(localDiff);
	  // Add on to position
	  GleemV3f::add(worldDiff, position, position);
	}
      else if (button1Down && button2Down)
	{
	  // Dolly functionality
	  // Compute the local coordinate system's difference vector
	  GleemV3f localDiff(0, 0, dollySpeed * -1.0f * dy / 100.0f);
	  // Rotate this by camera's orientation
	  GleemV3f worldDiff = orientation.rotateVector(localDiff);
	  GleemV3f::add(worldDiff, position, position);
	  // Subtract off the dolly amount from the focal distance
	  float focalDiff = (float) worldDiff.length();
	  if (dy > 0)
	    focalDiff *= -1.0f;
	  focalDist += focalDiff;
	  if (focalDist < minFocalDist)
	    focalDist = minFocalDist;
	}
    }
}

void
ExaminerViewer::reshapeMethod(int w, int h)
{
  glViewport(0, 0, w, h);
  float aspect, theta;
  aspect = (float) w / (float) h;
  if (w >= h)
    theta = 45;
  else
    theta = GLEEM_RAD_TO_DEG(atan(1 / aspect));
  params.vertFOV = GLEEM_DEG_TO_RAD(theta) / 2.0f;
  params.imagePlaneAspectRatio = aspect;
  params.xSize = w;
  params.ySize = h;
}

void
ExaminerViewer::recalc()
{
  makeCurrent();
  // Recompute position, forward and up vectors
  params.position = position;
  orientation.rotateVector(GleemV3f(0, 0, -1), params.forwardDirection);
  orientation.rotateVector(GleemV3f(0, 1, 0), params.upDirection);

  //  cerr << "params.position: " << params.position << endl;
  //  cerr << "params.forwardDirection: " << params.forwardDirection << endl;
  //  cerr << "params.upDirection: " << params.upDirection << endl;
  //  cerr << "focalDist: " << focalDist << endl;

  // Compute modelview matrix based on camera parameters, position and
  // orientation
  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  GleemV3f axis;
  float ang = orientation.getValue(axis);
  if (axis.lengthSquared() > EPSILON)
    glRotatef(GLEEM_RAD_TO_DEG(ang), -axis[0], -axis[1], -axis[2]);
  glTranslatef(-position[0], -position[1], -position[2]);
  //  glTranslatef(0, 0, 10);
  //  glRotatef(15, 0, 1, 0);
  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  gluPerspective(GLEEM_RAD_TO_DEG(params.vertFOV * 2.0f),
		 params.imagePlaneAspectRatio,
		 1, 100);
}

ExaminerViewer *
ExaminerViewer::getExaminerViewer(int windowID)
{
  WindowToExaminerViewerTable::iterator iter =
    windowViewerTable.find(windowID);
  if (iter == windowViewerTable.end())
    return NULL;
  return *iter;
}
